439 no support for buildx build build context

bigcat2014 opened the pull request:
Fixes #439

gabrieldemarmiesse commented:
Sorry it seems there is an issue with the CI. I'll try to fix it quickly. Thanks for your patience! In the meantime, could you add a simple unit test for this new feature?

gabrieldemarmiesse commented on python_on_whales/components/buildx/cli_wrapper.py:
```diff
@@ -208,7 +208,7 @@ def build(
         attest: Optional[Dict[str, str]] = None,
         build_args: Dict[str, str] = {},
-        # TODO: build_context
+        build_contexts: Dict[str, str] = {},
```
```suggestion
        build_contexts: Dict[str, Union[str, ValidPath]] = {},
```
Any path-like object will also be accepted for local directories

bigcat2014 commented:
Working on unit tests at the moment, although this is my first contribution to open source and I've never had to do unit tests before, so it's taking me a min to work through the structure of them. Should be able to have something to review soon though.

bigcat2014 commented:
So, with the current unit tests, 2 of the ones that I added are expected to fail until it is merged into master because they use absolute paths to files that exist on the branch, which will become master once this is merged (references a `test_context` directory that I created in the git repository and a .tar file)

bigcat2014 commented:
@gabrieldemarmiesse Any movement on this recently? Anything I can update in the unit tests to move this along more quickly?

gabrieldemarmiesse commented:
yeah I'm sorry, this pull request fell at the bottom of the list. Thanks for reminding me. It seems the tests are not trivial so I think I'll handle them myself. I'll likely have some time tomorrow afternoon to work on it. Thanks for your patience!

bigcat2014 commented:
I've updated some of the unit tests, adding a test for oci layout compliant directories and removing the test for .tar files as they're not explicitly supported by the --build-context flag. The only test that's still a little goofy to implement is the git repository one, but I think what I have is at least a good starting point for you!

gabrieldemarmiesse commented:
thanks I'll try it out :)

gabrieldemarmiesse commented:
@bigcat2014 do you need help?

bigcat2014 commented:
I'm not sure, I keep running the unit tests locally, they pass, then they don't pass the automated checks, so I'm not sure exactly what's going on. If the current ones don't pass then I may need you to take a look

bigcat2014 commented:
Ok, the unit tests I have added seem to be passing now. I think it was an issue with my understanding of the "with_docker_driver" and "with_container_driver" fixtures. I updated my tests to use what I think are the correct fixtures now and it seems to be working. I think I've also covered all of the --build-context use cases (local path, git repo, oci layout compliant directory, and docker image)

bigcat2014 commented:
Ok, I'm not entirely sure what's going on. I had all of the unit tests passing except the linting, I ran the linter to fix formatting. It only updated the test_buildx_cli_wrapper.py file and only updated formatting of the unit tests I added, but now a different unit test that I did not add, which was not reformatted at all, is failing.

gabrieldemarmiesse commented on tests/python_on_whales/components/buildx/test_buildx_cli_wrapper.py:
```diff
@@ -30,6 +46,25 @@ def with_container_driver():
+    docker.buildx.use(current_builder)
+
+    # Extract tar to directory
+    tar = tarfile.open(tar_path)
+    tar.extractall(oci_folder_path)
+    tar.close()
```
```suggestion
    with tarfile.open(tar_path) as tar:
        tar.extractall(oci_folder_path)
```

gabrieldemarmiesse commented:
We do have some flaky tests here and there, so it's possible that some tests fail randomly :(

gabrieldemarmiesse commented:
thanks a lot! It's going to help a lot of users :)
